# Introduction

Mainstream media is a traditional reference for news entities capable of shaping and molding opinions of the general public. The writers capable of reaching out to a different level of audiences. The dissemination of knowledge covers a wide variety of platforms, including television, online, articles, social media, and more.

The Limitless capabilities of communication allow news agencies, writers to maximize information delivery. Moreover, benefits potentially include a better-informed and engaged general public. However, the writer can express their feelings or utilizing statistics to avoid biases. The writer can analyze and write an article based on experience, observation, psychology, statistical reports, and feelings, and ideas.

This article that is going to instigate whether the writer has change in his writing approached. Choosing two articles written by Russell Carlton for the year 2009 and the year 2018 in the website Baseball Prospectos was he rely heavily on statistics and the psychological side of the report. 

One of the main themes in NLP is text representation, which is fundamental and indispensable for text-based intelligent information processing. Generally, text representation word frequencies examining the relevance of key-words to documents in the corpus.

# Text Preprocessing
We will begin loading the nltk and re packages that will allow us to process both texts and do some initial cleaning, such as lowercasing words and stemming.


We will first lowercase every word within each corpus. Then, we will remove stopwords from both corpora; however, we will not add any initial stopwords to nltk's stopwords list. If any additional words show up within the corpora that are deemed unnecessary, these words will then be added and will be reprocessed.
Finally, we will stem every token within the corpora.

In [49]:
# Loading required packages
import nltk
from nltk.corpus import stopwords
from nltk import bigrams
from nltk.collocations import *
import re

In [66]:
R09 = open("Russel1.txt", "r", encoding="latin-1")
R18 = open("Russel9.txt", "r", encoding="latin-1")

In [67]:

# Reading the texts
RussY09 = R09.read()
RussY18 = R18.read()

In [68]:
# Tokenize words to check length of corpora
RussY09_tokens = nltk.word_tokenize(RussY09)
RussY18_tokens = nltk.word_tokenize(RussY18)


In [69]:
# Print token length
print('RussY09: {} tokens'.format(len(RussY09_tokens)))
print('RussY18: {} tokens'.format(len(RussY18_tokens)))

RussY09: 1369 tokens
RussY18: 1650 tokens


In [70]:
# Create lowercase function that loops through every token in a corpus and returns a list of each lowercased token. 
def lowercase(text):
    token_list = []
    for word in text:
        word_lower = word.lower()
        token_list.append(word_lower)
    return(token_list)

# Lowercase tokens in each text for easier stop word removal, stemming, and do frequency distribution
# for each corpus. 
td_lower = lowercase(RussY09_tokens)
fw_lower = lowercase(RussY18_tokens)

In [71]:
# Remove stop words
stop_words = stopwords.words('english')
td_lower_stop = [word for word in td_lower if word not in stop_words]
fw_lower_stop = [word for word in fw_lower if word not in stop_words]

In [72]:
# Count the new number of tokens now that stopwords have been removed.
print('RussY09: {} tokens'.format(len(RS09_lower_stop)))
print('RussY18: {} tokens'.format(len(RS18_lower_stop)))

RussY09: 805 tokens
RussY18: 919 tokens


# After removing stop words
By removing stop words words drop down dramintcally 2009 articles drop 564 from 1369 to 805 tokens and 731 from 1650 to 919 tokens

In [73]:
# stem the words 
porter = nltk.PorterStemmer()

        
td_stemmed = []
for word in td_lower_stop:
    stemmed = porter.stem(word)
    td_stemmed.append(stemmed)

fw_stemmed = []
for word in fw_lower_stop:
    stemmed = porter.stem(word)
    fw_stemmed.append(stemmed)

In [74]:
# Remove special characters and print out the first 50 words for each corpus.
# The RussY09 for the year 2019
td_clean = [re.sub(r'[^a-zA-Z0-9]+', '', word) for word in td_stemmed]
td_clean = list(filter(None, td_clean))

#RussY09 for the year 2018
fw_clean = [re.sub(r'[^a-zA-Z0-9]+', '', word) for word in fw_stemmed]
fw_clean = list(filter(None, fw_clean))


In [75]:
# Use NLTK's FreqDist function to calculate token counts on the cleaned corpora
td_freq = nltk.FreqDist(td_clean)
fw_freq = nltk.FreqDist(fw_clean)

# Print the 50 most common tokens in The Dead...
print("The 50 most common words in The Russel_Year_09 are:\n")
print(td_freq.most_common(50))

The 50 most common words in The Russel_Year_09 are:

[('jay', 16), ('make', 16), ('halladay', 15), ('team', 14), ('blue', 13), ('trade', 12), ('better', 10), ('get', 10), ('offer', 8), ('roy', 7), ('would', 7), ('valu', 7), ('two', 7), ('say', 7), ('go', 6), ('game', 6), ('propos', 6), ('it', 6), ('like', 6), ('deal', 5), ('prospect', 5), ('he', 5), ('pick', 5), ('three', 5), ('year', 4), ('market', 4), ('away', 4), ('noth', 4), ('decis', 4), ('accept', 4), ('even', 4), ('fair', 4), ('doesnt', 4), ('guy', 4), ('ye', 4), ('bid', 4), ('basebal', 3), ('ultimatum', 3), ('toronto', 3), ('young', 3), ('name', 3), ('packag', 3), ('alex', 3), ('anthopol', 3), ('pitcher', 3), ('free', 3), ('ed', 3), ('walk', 3), ('probabl', 3), ('nickel', 3)]


In [76]:
# ...and the 50 most common in Russel_Year_18. 
print("The 50 most common words in Russel_Year_18 Wake are:\n")
print(fw_freq.most_common(50))
n_print =50

The 50 most common words in Russel_Year_18 Wake are:

[('ball', 32), ('fielder', 23), ('left', 18), ('get', 12), ('might', 9), ('outfield', 8), ('batter', 8), ('good', 8), ('defens', 7), ('singl', 7), ('go', 7), ('worth', 7), ('percent', 7), ('one', 6), ('base', 6), ('need', 6), ('wall', 6), ('hit', 6), ('doubl', 6), ('run', 6), ('thing', 5), ('talk', 5), ('cut', 5), ('infield', 5), ('dont', 5), ('make', 5), ('time', 5), ('play', 5), ('way', 5), ('could', 5), ('valu', 5), ('field', 5), ('wer', 5), ('data', 5), ('300', 5), ('bad', 5), ('want', 4), ('it', 4), ('there', 4), ('gap', 4), ('stop', 4), ('fact', 4), ('pick', 4), ('take', 4), ('littl', 4), ('know', 4), ('would', 4), ('case', 4), ('extra', 4), ('ground', 4)]


# Most common words analysis

The 50 words frequency in the year 2009 is all about the team is seems blue jay appears(208) times and Halladay(189) is the main topic about prospect, proposal, and offer. The offer was not fear and someone walkaway. 

On the other hand, in 2018, it is a different flavor about the skills; the ball appears 349 times left fielder; these always go together in baseball left fielder scores better than the centerfielder. Talking about the batter, the hit, worth, percentage, and also talks about the defense and all components in the baseball game.  

In [77]:
# Create the bigram finder and score bigram frequency
# We'll first work with The Russel_Year_09
bigram_measures = nltk.collocations.BigramAssocMeasures()
td_finder = BigramCollocationFinder.from_words(td_clean)
td_scored = td_finder.score_ngrams(bigram_measures.raw_freq)

In [78]:

# Print the first 50 bigram scores for  The Russel_Year_09
for bscore in td_scored[:50]:
    print (bscore)

(('blue', 'jay'), 0.016516516516516516)
(('roy', 'halladay'), 0.009009009009009009)
(('accept', 'propos'), 0.0045045045045045045)
(('alex', 'anthopol'), 0.0045045045045045045)
(('draft', 'pick'), 0.0045045045045045045)
(('make', 'better'), 0.0045045045045045045)
(('make', 'decis'), 0.0045045045045045045)
(('say', 'ye'), 0.0045045045045045045)
(('two', 'three'), 0.0045045045045045045)
(('walk', 'away'), 0.0045045045045045045)
(('best', 'offer'), 0.003003003003003003)
(('blue', 'jays'), 0.003003003003003003)
(('cy', 'young'), 0.003003003003003003)
(('decis', 'make'), 0.003003003003003003)
(('ed', 'scarborough'), 0.003003003003003003)
(('firstround', 'draft'), 0.003003003003003003)
(('he', 'go'), 0.003003003003003003)
(('jay', 'get'), 0.003003003003003003)
(('jay', 'would'), 0.003003003003003003)
(('team', 'bid'), 0.003003003003003003)
(('three', 'year'), 0.003003003003003003)
(('ultimatum', 'game'), 0.003003003003003003)
(('valu', 'player'), 0.003003003003003003)
(('would', 'accept'), 0.

In [79]:
# Print the first 50 bigram scores for Russel_Year_2018
fw_finder = BigramCollocationFinder.from_words(fw_clean)
fw_scored = fw_finder.score_ngrams(bigram_measures.raw_freq)

# Print the first 50 bigram scores for Russel_Year_2018
for bscore in fw_scored[:50]:
    print (bscore)

(('left', 'fielder'), 0.023004059539918808)
(('cut', 'ball'), 0.005412719891745603)
(('make', 'play'), 0.005412719891745603)
(('1520', 'percent'), 0.0040595399188092015)
(('ball', 'get'), 0.0040595399188092015)
(('ball', 'hit'), 0.0040595399188092015)
(('ball', 'travel'), 0.0040595399188092015)
(('wer', 'talk'), 0.0040595399188092015)
(('10point', 'gap'), 0.0027063599458728013)
(('ball', 'certainli'), 0.0027063599458728013)
(('ball', 'field'), 0.0027063599458728013)
(('center', 'fielder'), 0.0027063599458728013)
(('could', 'make'), 0.0027063599458728013)
(('data', 'set'), 0.0027063599458728013)
(('defens', 'metric'), 0.0027063599458728013)
(('defens', 'worth'), 0.0027063599458728013)
(('fact', 'batter'), 0.0027063599458728013)
(('field', 'outfield'), 0.0027063599458728013)
(('fielder', 'make'), 0.0027063599458728013)
(('front', 'left'), 0.0027063599458728013)
(('get', 'wall'), 0.0027063599458728013)
(('go', 'way'), 0.0027063599458728013)
(('hit', 'ground'), 0.0027063599458728013)
(('kn

# Bigram Analysis
The bigrams are very interesting how words connect each other, although blue and jay appeared many times, to begin with still in the top one with the word frequency of 0.017% of likelihood ratio and raw frequency. Next is Roy, and the Hallway, machine knew that these words are pair together. Upon looking with the pair of words,  is confirmed the topic between Blue Jays and Roy Hallaway, the writer(Russell) talking if Hallaway will pick two or three persons and seems blue jays walk away from the offer. The writer continued to talk about the blue jay. One that is very noticeable the word better describes in so many ways but has the lowest word frequency.

The year 2018 talking about a baseball game norm was the left fielder with a bigram score of 0.023%.  And who make to the about the playoff. Upon looking at 1520, it puzzled me, but later I  figure it out it is 15-20 percent, during the process it got deleted the dash. Base on this data, the writer uses statistics to perform the analysis. One that is very noticeable the word ball describes in so many ways. Besides that, the writer is very connected to the event. 

# Pointwise Mutual Information (PMI)

In [80]:
# Apply a mininmum word frequency of five to The year 2009 on the top 50 bigrams. 
td_finder.apply_freq_filter(5)
td_scored = td_finder.score_ngrams(bigram_measures.pmi)
for bscore in td_scored[:50]:
    print (bscore)

(('roy', 'halladay'), 5.250095350126295)
(('blue', 'jay'), 5.138370267567467)


In [81]:
# Repeat the same exercise for the year 2018
fw_finder.apply_freq_filter(5)
fw_scored = fw_finder.score_ngrams(bigram_measures.pmi)
for bscore in fw_scored[:50]:
    print (bscore)

(('left', 'fielder'), 4.923406437897166)


# PMI Analysis
Surprisingly the PMI is very good. Only one word that don't occur together the world 'littil' for the word little in the 2009 article.  All are good collocation pairs have high PMI because the probability of co-occurrence is only slightly lower than the probabilities of occurrence of each word. Conversely, a pair of words whose probabilities of occurrence are considerably higher than their likelihood of co-occurrence gets a small PMI score.

Next, using PMI setting a five-word frequency limit. Both articles have 50 qualifying bigrams. These all refer to characters and actions associated with the characters. 

In the 2009 article, Russell uses a character-centered approach, focusing on the situation itself and how characters react to the same events.

On the other hand, the 2018 article. Action-based writing. The author keeps the action flowing.  There are many descriptive bigrams that either involve adjectives, nouns, or adverbs.

# The Two Articles
The two articles have a different flavor of the year, month, or day. The author is very focused on both topics. 

The top 50 words very much are the summary of the story. Next is the value of each word frequency very much the same.

Base on two articles, the author changes his writing style base on the event and maintains the efficacy. Upon reading the top 50 words, the writer is has a few emotional peaks. 

Overall, there aren't many differences between both articles' bigrams. However, there is a difference between corpora, denoting how Russell's writing style evolves from the one story to the next. In 2009 about the controversy on Blue Jay and Halladay. That Blue Jay walks away from the offer, and in the second article is the NBA norm game. 

# Conclusion
Conducting analysis frequently requires skills and the wisdom to select effective methodologies. Without applying foundational principles for data preparation and associated analysis, could risk time management.

NLP is an exciting field for data scientists and engineers. Non-commercial packages and utilities allow coding challenges to be abstracted. Moreover, numerous off-shelf solutions require knowledge to select the right combination of frameworks. Often, the acumen in engineering a problem (or lack of), is noticeable within the associated results. Furthermore, in the area of text mining and NLP, the combination of tokenization and vectorization provide data scientists room for creativity